### DISCUSSION TITLE
Support for "reactive" Flow trigger type

### DISCUSSION THREAD

**Discussion by emahuni** [directus/directus#18886] [Feature Requests]
### Summary

I expected Directus to have an event type that uses **final** event data after an event has **completed** successfully. I tried both event trigger types `filter` and `action`, and noticed they serve their intended purposes very well, and lack a way to work with final resulting event data especially when working with `items.create`, `items.update` and `item.delete` (sorry to digress; there should also be an `items.read` and `items.count`). In addition to `filter` and `action` types there could be another missing type that we may call `reactive`.

![image](https://github.com/directus/directus/assets/7290508/e28d04ab-18da-466b-b253-f25867c29a05)

 We want to be able to react only when the event was successfully completed. The operations in the triggered flow will work with the resulting final data. In this type, we won't care about aborting or cancelling the event as it would have already successfully happened. It's like an **after** hook, considering the **filter** and **action** types are **before** hooks.

More like, if we successfully created an item in collection A, then use the resulting item data to do blah blah operation.

### Basic Example

For example, I have a Menu collection that has foods collection related to it. A user can create additional foods items whilst interacting with a menu item. When the user saves the Menu, an event hook flow can be triggered that requires all foods items' IDs to be used to notify a user that the Menu has **these** new food items, of which user can directly interact with the food items based on their IDs. The flow is only triggered **AFTER** the Menu item and all its foods items have been saved.

This cannot be achieved using the filter type as it only works with the data in its pre-saved phase. The Action type is also similar but is just sort of the async version of filter. Therefore the current implementations will provide unsaved foods items and existing foods items in an array without a way to act on the resulting final data.

### Motivation

I am coming from `Strapi` and there are `before` and `after` hooks in the API where this can be defined easily. They allow dev to modify or cancel an update, creation etc in the before hook or modify or do something else after an update, creation, etc in after hooks. see the following: https://docs.strapi.io/dev-docs/backend-customization/models#lifecycle-hooks

There are tonnes of reasons why this is essential and having worked with it before kind of makes it strange that Directus doesn't support it.

### Detailed Design

I looked into the src code and noticed that events are elegantly handled. This could be done here:

![image](https://github.com/directus/directus/assets/7290508/0eb0be2c-bdfb-4683-b316-1d294e778a7d)

just before returning the resulting data, somewhere there, another emission such as in  can be done to invoke the intended hooks. This should be done on all returns above properly.

or here:

![image](https://github.com/directus/directus/assets/7290508/e90b106c-2f09-409c-8aa5-fae0f5f9a09d)

The idea is to trigger other flows or operations just after everything is done.

### Requirements List

___

### Drawbacks

The feature is a core api implementation and requires modification of the core. The modification costs are very little as there just need to be non-blocking events emitted at the correct point of an event. 

### Alternatives

I thought of using the action event type and use several flows to save any unsaved items manually before  passing the data to another flow to handle (since one can't loop in operations) the data the way it should be handled in the after event hook. However this introduces a lot of complexity and results in blotted and unreliable logic.

### Adoption Strategy

existing Directus devs can just adopt it without any breaking changes or migrations. It won't affect any other projects as it will be up to a dev to use it or not.

### Unresolved Questions

Not yet sure **exactly** where to put the event and how the flow triggering system works. Once I figure it out will also update here. I am open to help implement it.

**Comment by rijkvanzanten** [comment]
The `action` should be what you're asking for here 🤔 It fires after the operation is successfully completed!

**Reply by emahuni** [reply]
Is it 😅? That's what I expected at first, but didn't get the intended effect and read back the docs, which actually made things worse. Then I guess documentation needs to be clearer. Documentation sound as if actions are just Async versions of filters.

"An Action lets the event commit its transaction to the database **without waiting for the flow to finish** _as if, docs is saying, flow starts as soon as event is emitted and nothing is connected to anything anymore, it doesn't wait for event._ Therefore, Actions give you a more performant API, but you have no ability to validate or transform the payload before the transaction is committed to the database. **However, a copy of the payload is still added into $trigger to use as desired.** _no info that this is the final result of the event passed to the action as payload_
"
It's as if action flows just go right ahead and work with the initial event payload (which is what I got when looking at $trigger.payload) without regard to the event progress the moment the event starts. This doesn't say anything about the final data, how to access it, or that an action flow waits for its event to complete and then triggers.

 I tried using it and didn't get the data I was expecting, didn't know if this was happening before or after the event. Maybe there's a bug somewhere or I was missing something along the way since the docs are not clear on this. Let me look into it again and revert back. Thanks.

**Comment by emahuni** [comment]
Okay, so I have tested it and it seems like there is no difference between the two in regards to the data they provide. Which led to my earlier assumptions. Below are pieces of output data that illustrate this. A collection called unit has self relating field `equates_to_unit` that simply points to another `unit` it is equivalent to.

Set up 2 flows, as follows:
for filter:
![image](https://github.com/directus/directus/assets/7290508/3c9e967c-8a8d-4fac-9dcb-ef6114a943b1)

for action:
![image](https://github.com/directus/directus/assets/7290508/3bbc5533-f137-4b23-8a87-ccda3bd029f2)

Now create a new unit call it litre and save it as follows:
![image](https://github.com/directus/directus/assets/7290508/ee54856d-0a0e-49c7-a42b-153db539a632)

Edit it and add a new unit in equates to as follows:
![image](https://github.com/directus/directus/assets/7290508/24809570-25ad-44a3-850a-2f3820d2ae1a)
![image](https://github.com/directus/directus/assets/7290508/797b65f4-9299-4971-8334-e61b8dc55b58)
and save both.

Below is the resulting data in the flows:
```js
// test filter - data:
{
  '$trigger': {
    payload: {
      equates_to_unit: { name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
      equates_to_qty: 1000
    },
    event: 'units.items.update',
    keys: [ '4', [length]: 1 ],
    collection: 'units'
  },
  // ... other data
}

// test action - data:
{
  '$trigger': {
    event: 'units.items.update',
    payload: {
      equates_to_unit: { name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
      equates_to_qty: 1000
    },
    keys: [ '4', [length]: 1 ],
    collection: 'units'
  },
  // ... other data
}

// test action - data:
{
  '$trigger': {
    event: 'units.items.create',
    payload: { name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
    key: 16,
    collection: 'units'
  },
}
```

As you can see, the filter and action data is the same. Though what I am saying may seem pointless at first, hear me out. With the filter, that's fine, it's usable data, but with the action there is no way to use `equates_to_unit` item without writing some funky code to figure out the actual item that was created (there is no ID/key) in the DB. The data looks exactly the same because it is. It seems like it creates the other `equates_to_unit` new unit later, after the action has fired. This prevents it from getting the actual final state of the `litre` unit in its action. Even if I wanted to read the data manually and figure out the new nested item in the action, it won't work since the root action happens before the nested item is created.

What I expected to see was something like this:

```js
// test filter - data:
{
  '$trigger': {
    payload: {
      equates_to_unit: { name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
      equates_to_qty: 1000
    },
    event: 'units.items.update',
    keys: [ '4', [length]: 1 ],
    collection: 'units'
  },
  // ... other data
}

// test action (this is creation of the nested item first before running the update action) - data:
{
  '$trigger': {
    event: 'units.items.create',
    payload: { name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
    key: 16,
    collection: 'units'
  },
}

// test action (triggering of the update action, nb it has an id added to equates_to_unit...
// this is what I am actually interested in and makes the data useful and accurate:. the item and its nested items were updated) - data:
{
  '$trigger': {
    event: 'units.items.update',
    payload: {
      equates_to_unit: { id: 16, name: 'Kilolitres', abbrv: 'kl', form_factor: 'liquid' },
      equates_to_qty: 1000
    },
    keys: [ '4', [length]: 1 ],
    collection: 'units'
  },
  // ... other data
}
```

Notice that the nested (new) 'kilolitre' unit item was created before the parent 'litre' action was run. The 'litre' action actually happened after everything was done, and the resulting data is sort of the final state of the 'litre' unit payload with the **id** of the nested item.

The ideal scenario would be for final **fields** to be specified in the action config, what should be given to the action. If none are specified then it gives back the above format, ie: the initial payload plus id(s) of the updated or new nested items. EG:

![image](https://github.com/directus/directus/assets/7290508/aa813845-f28d-4306-8e89-5e4b683f610d)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
